### PR TITLE
feat: storage: Path type filters

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Implements https://github.com/filecoin-project/lotus/issues/8794

## Proposed Changes
Add two new fields to `[storagepath]/sectorstore.json`:
```yaml
// AllowTypes lists sector file types which are allowed to be put into this
// path. If empty, all file types are allowed.
//
// Valid values
// - "unsealed"
// - "sealed"
// - "cache"
// - "update"
// - "update-cache"
// Any other value will generate a warning and be ignored.

"AllowTypes": [
  "sealed", "unsealed", "update",
],



// DenyTypes lists sector file types which aren't allowed to be put into this
// path.
//
// Valid values
// - "unsealed"
// - "sealed"
// - "cache"
// - "update"
// - "update-cache"
// Any other value will generate a warning and be ignored.

"DenyTypes": [
  "cache", "update-cache"
]
```

After a path with this config is attached, lotus-miner will stop putting new files of disallowed types into that path.

If there are existing files with disallowed types in a path, those files will remain readable for PoSt/Retrieval, so the worst that can happen in case of misconfiguration is that sealing tasks will get stuck waiting for storage to become available.

The most basic setup to separate sealed data from unsealed copies would be to:
* Not have any filters on sealing paths (no change)
* Add `"DenyTypes": ["unsealed"]` to long-term sealed storage paths
* Add `"AllowTypes": ["unsealed"]` to long-term unsealed storage paths

Here's how storage list looks with filters set:
2022-07-12-131643_816x445_scrot

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

File: itests/kit/blockminer.go [hunk 1]
```diff
@@ -184,7 +185,12 @@ func (bm *BlockMiner) MineBlocksMustPost(ctx context.Context, blocktime time.Dur
 
 			var target abi.ChainEpoch
 			reportSuccessFn := func(success bool, epoch abi.ChainEpoch, err error) {
-				require.NoError(bm.t, err)
+				// if api shuts down before mining, we may get an error which we should probably just ignore
+				// (fixing it will require rewriting most of the mining loop)
+				if err != nil && !strings.Contains(err.Error(), "websocket connection closed") {
+					require.NoError(bm.t, err)
+				}
+
```
Comment (magik6k) [inline itests/kit/blockminer.go:193]: review note: This isn't exactly related to this PR, but it should fix lots of test flakiness we're seeing

File: storage/sealer/storiface/filetype_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	// allow denies non-matching types
+	require.False(t, FTUnsealed.Allowed((FTCache).Strings(), nil))
+}
+
+func TestFileTypeAnyAllow(t *testing.T) {
+	// no filters = allow all
+	require.True(t, FTCache.AnyAllowed(nil, nil))
+
+	// one denied
```
Comment (arajasek) [inline storage/sealer/storiface/filetype_test.go:33]: Can you add 

```
require.True(t, FTCache.AnyAllowed(nil, (FTUnsealed).Strings()))
```

File: storage/paths/index.go [hunk 3]
```diff
@@ -107,6 +116,64 @@ func (i *Index) StorageList(ctx context.Context) (map[storiface.ID][]storiface.D
...
+			}
+
+			continue
+		}
+		allow = append(allow, typ)
+	}
+	for id, typ := range si.DenyTypes {
+		_, err := storiface.TypeFromString(typ)
+		if err != nil {
+			// No need no hard-fail here, just warn the user
```
Comment (arajasek) [inline storage/paths/index.go:151]:
```suggestion
			// No need to hard-fail here, just warn the user
```

**Review by arajasek — APPROVED** [review]
Nits, but looks like it works.

File: storage/paths/index.go [hunk 4]
```diff
@@ -107,6 +116,64 @@ func (i *Index) StorageList(ctx context.Context) (map[storiface.ID][]storiface.D
 }
 
 func (i *Index) StorageAttach(ctx context.Context, si storiface.StorageInfo, st fsutil.FsStat) error {
+	var allow, deny = make([]string, 0, len(si.AllowTypes)), make([]string, 0, len(si.DenyTypes))
+
+	if _, hasAlert := i.pathAlerts[si.ID]; i.alerting != nil && !hasAlert {
+		i.pathAlerts[si.ID] = i.alerting.AddAlertType("sector-index", "pathconf-"+string(si.ID))
+	}
+
+	var hasConfigIsses bool
```
[the remark this answers is not in the document]
Comment (magik6k) [reply storage/paths/index.go:125]: It tells you whether there are issues in the storage path config.

**Pull request merged**